### PR TITLE
WIP: First proposal of real-time notification system

### PR DESCRIPTION
**Pull request by jvican**
Flask-notifications is a Flask extension that adds support for real-time notifications,
with a simple API that allows a good control over different consumers (e.g. email or push notifications).

I am a CERN Openlab Summer Student developing this system for the Invenio team. This is a first
draft of the system, which process the notification delivery in an asynchronous way thanks to Celery.
I use Redis as a broker -this could eventually change- and Flask-Mail to send emails. It includes support for Push notifications (SSE).

It's not finished.

### CODE REVIEW

File: .kwalitee.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+components:
+- README
+- docker
+- docs
+- kwalitee
+- menu
```
Comment (jirikuncar) [inline .kwalitee.yml]: notifications

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Flask-Menu
```
Comment (jirikuncar) [inline .travis.yml]: replace Flask-Menu with Flask-Notifications

File: AUTHORS [hunk 3]
```diff
@@ -0,0 +1,15 @@
+Authors
+-------
+
+Flask-Menu is developed for use in `Invenio <http://invenio-software.org>`_
+digital library software.
+
+Contact us at `info@invenio-software.org <mailto:info@invenio-software.org>`_
+
+* Krzysztof Lis <krzysztof.lis@cern.ch>
```
Comment (jirikuncar) [inline AUTHORS]: you should be the only author so far

File: AUTHORS [hunk 4]
```diff
@@ -0,0 +1,15 @@
+Authors
+-------
+
+Flask-Menu is developed for use in `Invenio <http://invenio-software.org>`_
```
Comment (jirikuncar) [inline AUTHORS]: `s/Flask-Menu/Flask-Notifications/`

File: CHANGES [hunk 5]
```diff
@@ -0,0 +1,23 @@
+Changes
+=======
+
+Version 0.3.0 (released 2015-03-17)
```
Comment (jirikuncar) [inline CHANGES]: keep only `Version 0.1.0`

File: CHANGES [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+- The Flask-Menu extension is now released under more permissive
+  Revised BSD License. (#12)
+- New support for additional keyword arguments stored as `MenuItem`
+  attributes. (#19)
+- Richer quick-start usage example. (#16)
+- Support for Python 3.4. (#6)
+- Documentation improvements. (#3)
+
+Version 0.1.0 (released 2014-06-27)
```
Comment (jirikuncar) [inline CHANGES]: `s/2014-06-27/TBD/` (To Be Defined)

File: CONTRIBUTING.rst [hunk 7]
```diff
@@ -0,0 +1,19 @@
+Contributing
+============
+
+Bug reports, feature requests, and other contributions are welcome.
+If you find a demonstrable problem that is caused by the code of this
+library, please:
+
+1. Search for `already reported problems
+   <https://github.com/inveniosoftware/flask-menu/issues>`_.
```
Comment (jirikuncar) [inline CONTRIBUTING.rst]: `s/flask-menu/flask-notifications/`

File: LICENSE [hunk 8]
```diff
@@ -0,0 +1,38 @@
+Flask-Menu is free software; you can redistribute it and/or modify it
+under the terms of the Revised BSD License quoted below.
+
+Copyright (C) 2013, 2014 CERN.
```
Comment (jirikuncar) [inline LICENSE]: keep only current year - 2015

File: README.rst [hunk 9]
```diff
@@ -1,48 +1,30 @@
-..
```
Comment (jirikuncar) [inline README.rst:1]: keep the original `README.rst`

File: RELEASE-NOTES.rst [hunk 10]
```diff
@@ -0,0 +1,47 @@
+===============================
```
Comment (jirikuncar) [inline RELEASE-NOTES.rst]: you can delete this file for the moment

File: docs/_templates/sidebarintro.html [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<h3>About</h3>
+<p>
+  <strong>Flask-Menu</strong> is an extension for Flask that adds support for quickly.
```
Comment (jirikuncar) [inline docs/_templates/sidebarintro.html]: `s/Menu/Notifications/`

File: docs/_templates/sse.html [hunk 12]
```diff
@@ -0,0 +1,30 @@
+<!DOCTYPE html>
```
Comment (jirikuncar) [inline docs/_templates/sse.html]: is it really a documentation template? ~~maybe you should move it to `examples/sse/templates/index.html`~~

File: tests/test_events.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+from flask import Flask
+from flask.json import jsonify, dumps
+from flask.ext.notifications import NotificationService, Event
+
+
+class EventsTest(unittest.TestCase):
+    def setUp(self):
+        self.app = Flask(__name__)
+
+        self.config = {
```
Comment (jirikuncar) [inline tests/test_events.py:47]: you should not keep the configuration here. you can use `console` email for testing.

Comment (jvican) [inline tests/test_events.py:47]: Sorry, what do you mean by console? I was thinking about just importing config from the main app.

File: tests/test_events.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
+import unittest
+from celery import Celery
+from flask import Flask
+from flask.json import jsonify, dumps
+from flask.ext.notifications import NotificationService, Event
```
Comment (jirikuncar) [inline tests/test_events.py]: Please use `from flask_notifications import ...`. `flask.ext` import hook is going to be deprecated in Flask 1.0. Thanks

File: tests/test_events.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
+import unittest
```
Comment (jirikuncar) [inline tests/test_events.py]: include the licence header in all files

File: setup.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        'Framework :: Flask',
+        'Topic :: Internet :: WWW/HTTP :: Dynamic Content',
+        'Topic :: Software Development :: Libraries :: Python Modules',
+        'Programming Language :: Python :: 2',
+        'Programming Language :: Python :: 2.6',
+        'Programming Language :: Python :: 2.7',
+        'Programming Language :: Python :: 3',
+        'Programming Language :: Python :: 3.3',
+        'Programming Language :: Python :: 3.4',
+        'Development Status :: 5 - Production/Stable'
```
Comment (jirikuncar) [inline setup.py]: maybe we should change the classifier to `'Development Status :: 1 - Planning',`

File: CHANGES (hunk 6 above)
Comment (tiborsimko) [inline CHANGES]: Cosmetics; `FIXME` is more usual than `TBD`, e.g. besides editor warnings I'm grepping for this from time to time.

File: setup.py (hunk 16 above)
Comment (tiborsimko) [inline setup.py]: :+1: let's keep updating the status as the package progresses

File: docs/index.rst [hunk 17]
```diff
@@ -0,0 +1,332 @@
+====================
```
Comment (jirikuncar) [inline docs/index.rst:10]: license

File: flask_notifications/consumers/__init__.py [hunk 18]
```diff
@@ -0,0 +1,9 @@
+from flask_notifications.consumers.push.push_consumer import PushConsumer
```
Comment (jirikuncar) [inline flask_notifications/consumers/__init__.py]: license

File: flask_notifications/consumers/__init__.py [hunk 19]
```diff
@@ -0,0 +1,9 @@
...
+from flask_notifications.consumers.log.log_consumer import LogConsumer
```
Comment (jirikuncar) [inline flask_notifications/consumers/__init__.py]: remove imports

File: flask_notifications/consumers/consumer.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
+#
```
Comment (jirikuncar) [inline flask_notifications/consumers/consumer.py:2]:
```
# -*- coding: utf-8 -*-
#
# This file ...
```

File: flask_notifications/consumers/log/.ropeproject/config.py [hunk 21]
```diff
@@ -0,0 +1,85 @@
+# The default ``config.py``
```
Comment (jirikuncar) [inline flask_notifications/consumers/log/.ropeproject/config.py]: remove this file

File: setup.py [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+        self.pytest_args = config.get("pytest", "addopts").split(" ")
+
+    def finalize_options(self):
+        TestCommand.finalize_options(self)
+        self.test_args = []
+        self.test_suite = True
+
+    def run_tests(self):
+        # import here, cause outside the eggs aren't loaded
+        import pytest
```
Comment (jirikuncar) [inline setup.py:38]: https://github.com/inveniosoftware/invenio-oauthclient/blob/master/setup.py#L57-L88

File: setup.py [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+    'flask-email>=1.4.4',
+    'flask-mail>=0.9.1'
+]
+
+setup(
+    name='Flask-Notifications',
+    version=version,
+    url='https://github.com/inveniosoftware/flask-notifications',
+    license='BSD',
+    author='jvican',
```
Comment (jirikuncar) [inline setup.py]: https://github.com/inveniosoftware/invenio-oauthclient/blob/master/setup.py#L103-L104

**Pull request merged**